### ISSUE TITLE
[Bug] Query user after login always returns `null`

### ISSUE THREAD

**Issue by sophiabrandt** [sophiabrandt/nextjs-ecommerce#1]
In production the `CurrentUserQuery` always returns `null`.
On logging in the return data from the Sign In mutation is set in the cache as the current user.

But as soon as you refetch queries, the current user gets overwitten with `null` again.

This is probably not an issue with Apollo cache, as the cache correctly shows the `authenticatedItem` with a `__ref` to the logged in user.

The issue only occurs in production with Heroku as Graphql API and Vercel as Next.js frontend. I wasn't able to reproduce it on my local environment, not even in production mode.

**Comment by sophiabrandt** [comment]
**tl;dr: this is an issue with how cookie authentication works.**

The backend (Keystone-Next) uses cookies for session management. When logging in, the backend sends a cookie to the browser.

Apollo should pick it up and pass it to the client for future browser HTTP requests (see options for [createUploadLink](https://github.com/jaydenseric/apollo-upload-client) in [`frontend/lib/apollo.ts`](https://github.com/sophiabrandt/nextjs-ecommerce/blob/main/frontend/lib/apollo.ts)).

However, the browser blocks the cookie that comes from the backend _if you use two different domains_ (cross-site requests). The Keystone session cookie is a third-party cookie in this setup.  
The (Keystone) cookie settings don't specify a `SameSite` attribute. Newer browser versions enforce `SameSite=Lax` in that case.

Keystone-Next does not allow you to modify the `SameSite` policy.

That means that  
a) the frontend (Next.js client) rejects the cookie sent from the backend  
b) even if you could set the cookie, you won't be able to send it to future POST requests to the GraphQL backend. `SameSite=Lax` doesn't send outgoing cookies with "unsafe HTTP methods" (POST method, needed for GraphQL)

**If you deploy to two different domains, you will need `SameSite=None` and `secure` for cookie settings.** This is currently not possible with Keystone-Next.

Possible workaround: deploy to the same domain to avoid cross-site resource requests.

Further Reading:

* https://next.keystonejs.com/apis/session#
* https://blog.heroku.com/chrome-changes-samesite-cookie
* https://www.valentinog.com/blog/cookies
* https://web.dev/samesite-cookies-explained/

**Comment by sophiabrandt** [comment]
Closed via deployment to a VPS.
